src/runs.c

```diff
@@ -0,0 +1,371 @@
+                                                               \
+  for (R_len_t i = 1; i < size; ++i) {                         \
+    const CTYPE elt = p_x[i];                                  \
+                                                               \
+    if (SCALAR_EQUAL(&elt, &ref) == 0) {                       \
+      ++id;                                                    \
+      ref = elt;                                               \
+    }                                                          \
+                                                               \
+    p_out[i] = id;                                             \
+  }                                                            \
+                                                               \
+  return id;                                                   \
+}
+
+static int lgl_identify_runs(SEXP x, R_len_t size, int* p_out) {
+  VEC_IDENTIFY_RUNS(int, LOGICAL_RO, lgl_equal_scalar_na_equal);
+}
+static int int_identify_runs(SEXP x, R_len_t size, int* p_out) {
+  VEC_IDENTIFY_RUNS(int, INTEGER_RO, int_equal_scalar_na_equal);
+}
+static int dbl_identify_runs(SEXP x, R_len_t size, int* p_out) {
+  VEC_IDENTIFY_RUNS(double, REAL_RO, dbl_equal_scalar_na_equal);
+}
+static int cpl_identify_runs(SEXP x, R_len_t size, int* p_out) {
+  VEC_IDENTIFY_RUNS(Rcomplex, COMPLEX_RO, cpl_equal_scalar_na_equal);
+}
+static int chr_identify_runs(SEXP x, R_len_t size, int* p_out) {
+  VEC_IDENTIFY_RUNS(SEXP, STRING_PTR_RO, chr_equal_scalar_na_equal);
+}
+static int raw_identify_runs(SEXP x, R_len_t size, int* p_out) {
+  VEC_IDENTIFY_RUNS(Rbyte, RAW_RO, raw_equal_scalar_na_equal);
+}
+
+#undef VEC_IDENTIFY_RUNS
+
+#define VEC_IDENTIFY_RUNS_BARRIER(SCALAR_EQUAL) {              \
+  int id = 1;                                                  \
+                                                               \
+  /* Handle first case */                                      \
+  int loc = 0;                                                 \
+  p_out[0] = id;                                               \
+                                                               \
+  for (R_len_t i = 1; i < size; ++i) {                         \
+    if (SCALAR_EQUAL(x, i, x, loc) == 0) {                     \
+      ++id;                                                    \
+      loc = i;                                                 \
+    }                                                          \
+                                                               \
+    p_out[i] = id;                                             \
+  }                                                            \
+                                                               \
+  return id;                                                   \
+}
+
+static int list_identify_runs(SEXP x, R_len_t size, int* p_out) {
+  VEC_IDENTIFY_RUNS_BARRIER(list_equal_scalar_na_equal);
```

Comment:
I think this would require a few changes to `list_equal_scalar_na_equal()`, which is used in a few places so it could probably be its own PR.

If you could please go ahead and push that change to master then I could make the change in a follow up PR.